javaagent-core/src/test/java/org/hypertrace/agent/core/config/AgentConfigSupplierTest.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright The Hypertrace Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.hypertrace.agent.core.config;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+import org.hypertrace.agent.config.Config;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+final class AgentConfigSupplierTest {
+
+  private CountingAgentConfigSupplier inner;
+  private AgentConfigSupplier agentConfigSupplier;
+
+  @BeforeEach
+  void beforeEach() {
+    // GIVEN a supplier of AgentConfig that also counts how many time its been invoked
+    inner = new CountingAgentConfigSupplier();
+    // WHEN we decorate it with the AgentConfigSupplier
+    agentConfigSupplier = new AgentConfigSupplier(inner);
+    // VERIFY starts out reporting its been invoked 0 times
+    assertEquals(0, inner.getCount());
+  }
+
+  @Test
+  void testMemoizingAgentConfig() {
+    // WHEN we access the config
+    final Config.AgentConfig agentConfig = agentConfigSupplier.get();
+    // VERIFY it has been invoked once
+    assertEquals(1, inner.getCount());
+
+    // WHEN we access the config
+    final Config.AgentConfig secondConfig = agentConfigSupplier.get();
+    // VERIFY the inner supplier has still only been invoked once
+    assertEquals(1, inner.getCount());
+    // VERIFY the two configs are the same instance
+    assertSame(agentConfig, secondConfig);
+  }
+
+  @Test
+  void testResetAgentConfig() {
+    // WHEN we access the config
+    final Config.AgentConfig agentConfig = agentConfigSupplier.get();
+    // VERIFY it has been invoked once
+    assertEquals(1, inner.getCount());
+
+    // WHEN we reset the supplier
+    agentConfigSupplier.reset();
+    // AND WHEN we access the config again
+    final Config.AgentConfig secondConfig = agentConfigSupplier.get();
+
+    // VERIFY it has been invoked twice
+    assertEquals(2, inner.getCount());
+    // VERIFY the two configs are different instances
+    assertNotSame(agentConfig, secondConfig);
+  }
+
+  /**
+   * thread-safe {@link Config.AgentConfig} {@link Supplier} that also counts how many times it has
+   * been invoked
+   */
+  private static final class CountingAgentConfigSupplier implements Supplier<Config.AgentConfig> {
+
+    private final AtomicInteger count = new AtomicInteger(0);
```

Comment:
This `AtomicInteger` is probably overkill, but I left it in on the off chance someone wants to expand on this test case later. I originally wrote this test to try to reproduce the double-checked locking race condition that can occur by using a multi-threaded scheduled executor service to ask for the `AgentConfig` 10,000 times via separate callables all calling `get` on the same `AgentConfigSupplier` instance and verifying that the inner `CountingAgentConfigSupplier` was only invoked once. I wasn't able to reproduce the bug in that fashion, so I pulled out the test case and simply tested the supplier "the normal way". Since `CountingAgentConfigSupplier` is initialized `BeforeEach`, there isn't necessarily a need to have this class be thread-safe and we could simply make this an `int`.  Happy to back it out to avoid complexity or keep it in if we want to expand the test cases around this more.